crates/chain/src/keychain/txout_index.rs

```diff
@@ -803,50 +665,67 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     /// the `target_index` is in the hardened index range), this method will make a best-effort and
     /// reveal up to the last possible index.
     ///
-    /// This returns an iterator of newly revealed indices (alongside their scripts) and a
-    /// [`super::ChangeSet`], which reports updates to the latest revealed index. If no new script
+    /// This returns list of newly revealed indices (alongside their scripts) and a
+    /// [`ChangeSet`], which reports updates to the latest revealed index. If no new script
     /// pubkeys are revealed, then both of these will be empty.
     ///
     /// Returns None if the provided `keychain` doesn't exist.
     pub fn reveal_to_target(
         &mut self,
         keychain: &K,
         target_index: u32,
-    ) -> Option<(
-        SpkIterator<Descriptor<DescriptorPublicKey>>,
-        super::ChangeSet<K>,
-    )> {
-        let descriptor_id = self.keychains_to_descriptors.get(keychain)?.0;
-        self.reveal_to_target_with_id(descriptor_id, target_index)
+    ) -> Option<(Vec<Indexed<ScriptBuf>>, ChangeSet<K>)> {
+        let mut changeset = ChangeSet::default();
+        let mut spks: Vec<Indexed<ScriptBuf>> = vec![];
+        while let Some((i, new)) = self.next_index(keychain) {
+            if !new || i > target_index {
+                break;
+            }
+            match self.reveal_next_spk(keychain) {
+                Some(((i, spk), change)) => {
+                    spks.push((i, spk));
+                    changeset.append(change);
+                }
+                None => break,
+            }
+        }
+
+        Some((spks, changeset))
```

Comment:
The way it's written this will never return `None`, I think?

Reply:
What if the keychain does not exist?

Reply:
I think I might try and make an Error for the keychain not exisitng.

Reply:
(nope it is fine as it is -- and evan is right it will return `None` when there is not keychain).

Reply:
You could add a line to check `self.keychain_to_descriptor_id`

```diff
     pub fn reveal_to_target(
         &mut self,
         keychain: &K,
         target_index: u32,
     ) -> Option<(Vec<Indexed<ScriptBuf>>, ChangeSet<K>)> {
+        let _ = self.keychain_to_descriptor_id.get(keychain)?;
         let mut changeset = ChangeSet::default();
+
         let mut spks: Vec<Indexed<ScriptBuf>> = vec![];
         while let Some((i, new)) = self.next_index(keychain) {
             if !new || i > target_index {
                 break;
             }
             match self.reveal_next_spk(keychain) {
                 Some(((i, spk), change)) => {
                     spks.push((i, spk));
                     changeset.append(change);
                 }
                 None => break,
             }
         }
 
         Some((spks, changeset))
     }
```